### PR TITLE
File Transfare API

### PR DESCRIPTION
**Pull request by WolverinDEV**
Hello,
i added some file API stuff to your API. I hope you like it and accept my pull request :)

### CODE REVIEW

**Comment by TheHolyWaffle** [comment]
After a quick glance: Lombok cannot be introduced in this project.

**Comment by mohe2015** [comment]
I think it is not that difficult to remove Lombok.

**Comment by WolverinDEV** [comment]
Okey well i can remove it but why? And can i ceep the junit test classes?

**Comment by TheHolyWaffle** [comment]
It is yet another extra steps for devs to getting started with the
project.  The test files are also in the wrong location too. They should be
in a test folder next to main folder to follow conventions.  I dont have
time to thoroughly check your PR right now unfortunately as I am on
vacation.
Op 19-jul.-2016 02:00 schreef "WolverinDEV" <notifications@github.com>:

**Comment by Firedroide** [comment]
@TheHolyWaffle Same, I also don't have that much time. I'll try to look into this PR over the next few days, though.

Since the discussion & code review for this PR are looking to get quite lengthy, I'll probably keep it in a direct Gitter chat - if that's ok for everyone involved, of course.

**Comment by Firedroide** [comment]
I've finally found some time to do a proper code review.
Since I also haven't received a response to the private Gitter conversation, I'll post my review here:

# Structural problems

- Adding Project Lombok is not a possibility.  
    And even if, this addition should've been its own pull request.
- Adding NonNull annotations also doesn't adhere to the rest of the API  
    The problem is that these annotations don't do anything anyways, they only do something
    if a user has actually taken the time to set up their IDE in a way so it complains if you pass `null`
- It's File**Transfer**, not FileTransfare
    - The past form of `read` is also `read` and not `readed`
- IO happening inside TS3Api and TS3ApiAsync
    - These classes are not meant to do that work, they're only an interface to the rest of the classes
    - IO for file transfers should probably happen in its own class, similar to how commands are sent by calling
        `TS3Query.doCommand`, which in turn places a command in the command queue of QueryIO
- Swallowing of IOExceptions
    - Why are we not telling the users what actually went wrong?
    - "Oh, look at this, the file I downloaded is -4 bytes long. Let's allocate a new array with that length!"
    - On a similar note, try-with-resources should've been used instead of manually closing all streams / sockets.
      Most finally blocks in this PR can cause `NullPointerException`s or other unexpected results
- In the asynchronous API, getting exceptions and IO errors is reported to `CommandFuture` as a success
- Several numeric error codes  
    This is Java and not C, geez
- Why does `downloadFile` directly store all the data into a byte array?
    - This makes downloading a file store it in memory in its entirety.  
      Let's assume someone were to download a file that's several gigabytes in size.
      That could lead to an `OutOfMemoryError` very quickly
    - We should just pass on the `InputStream` and let the user do the writing
- Same goes for file uploads, passing in a `byte[]` might not always work.
    - We should also accept `OutputStream`s or just `File`s
- `downoadFile` and `uploadFile` in TS3ApiAsync spawn new, non-pooled Threads
    - We have our own thread pool for IO operations, which should've been used here instead.
    - Alternatively, Java also has its own asynchronous IO API, which should be looked into.
- File sizes should not be an `int` but rather a `long`.
    - Files on TS3 servers are not limited to 2.1 GB as far as I can tell
- Do `CFtInitDownload` and `CFtInitUpload` really need inheritance?
    - Moreover, the superclass `CFtInitAbstract` is not actually `abstract`!
- The class `FileList` serves no actual purpose and should be replaced with `List<TS3File>`
    - It extends `Wrapper` but doesn't actually use any of its funtionality
    - Filtering out one concrete file should be left to the user.
    - Java 8 added streams on collections which are prefect for these kinds of operations.
- `TS3File` should not be an inner class with several fields, but rather the actual public `Wrapper` class.
    - It should probably also be named differently, since no other classes in the `wrapper` package are prefixed with "TS3"
- FileList does not return the correct values for files which are still being updated.
- Tests are not actually unit tests, but integration tests
    - Moreover, the tests take around 1 minute to complete, which is unacceptable
    - Is there a specific reason why these tests need a different package name than the rest of the API?
    - Why on earth does `TestFileDownload` open a `JFrame`?
    - All of this defies the purpose of these tests: **Automated** testing.


# Specific code problems

- TS3Api l 3715, l 3756  
    `if(transfare.getPort() == -1)`  
    All negative port numbers and 0 should throw an exception, not just -1, and not just a different return value
- CFtInitAbstract l 36  
    `add(new KeyValueParam("cpw", channel == null ? "" : channel.get("CHANNEL_PASSWORD")));`  
    `channel.get("CHANNEL_PASSWORD")` won't work for 3 reasons:  
    - Channel properties are always lowercase. You should've used `ChannelProperty.CHANNEL_PASSWORD` instead.
    - `Channel` doesn't actually contain the channel password, only `ChannelInfo` does
    - The passwords are not actually sent back in clear text, but rather encrypted.
      Sending the encrypted password back to the server likely won't help you
- CFtInitUpload l 40  
    `add(new KeyValueParam("size", this.size = size));`  
    Hinders clarity. Assignment should be done on a separate line.
- FileTransfare l 38  
    `this.fileKey = get("ftkey").getBytes();`  
    This uses the platform's default charset, which might not do the same thing on all systems.
    You should always specify which charset to use, in this case UTF-8 from `StandardCharsets`
    `fileKey` should probably also be stored as a String and only be converted once it's actually needed as a byte array.
- FileList l 70  
    `if(file.getName().equals(name) || (ignoreCase && file.getName().equalsIgnoreCase(name)))`  
    If `ignoreCase` is true, this performs redundant work
- TS3Api l 3805  
    Unused private method `getFileBytes`, which is also duplicated in TestFileUpload
- TestFileList  
    Does not close the connection to the TS3 server.
- Calls to `new Integer(value)` in TS3ApiAsync
    - Since we're using Java 7 and not Java 4, auto-boxing is probably a good idea
    - This call also creates a new `Integer` instance for every call instead of using internally cached ones.


# Topics needing discussion

- Should several simultaneous connections to the file transfer "server" be possible?  
    Is that officially supported?  
    Have you done any tests to confirm that multiple up- and downloads at the same time are possible?
- Is there also flood protection in place for file transfers?  
    If that were the case, we'd have to make sure we'd never get ourselves banned due to flooding
- What exactly should we do if the length of the data in FileTransfer and the data for `uploadFile` don't match up?


# Code Style

- Lots of whitespace on otherwise empty lines
    - pom l 281
    - TS3Query l 201
    - TS3Api l 3692, l 3700, l 3704, ...
    - TS3ApiAsync l 4005, l 4009, l 4022, ...
    - TestFileList l 40, l 54
    - etc
- Trailing whitespace
    - CFtGetFilelist l 33
    - CFtInitAbstract l 36
- Whitespace around `(`, `)`, `{`, `}`, and `,`
    - TS3Api l 3689, l 3693, l 3695, l 3701, ..., l 3712, l 3713, ..., l 3720, l 3727, ...
    - TS3ApiAsync l 4006, l 4010, ...
    - etc
- Missing empty line at the end of the file
    - TestFileDownload l 65 / 66
- Lots of fields with only a `@Getter`, but they're not actually made final
    - Basically all wrapper files and CFtInitUpload
- Duplicate `uploadFile` and `downloadFile` in TS3Api and TS3ApiAsync
    Again showing that the API classes are not a good place to be doing IO operations
- Unneeded Imports
    - TS3Api l 52 `import java.nio.channels.NetworkChannel;`
- Capitalization
    - Class FileList vs CFtGetFilelist
- This might have just been a quirk of my current improv git install, but some of the files were showing up with an LF ending
    - The rest of the project is using CRLF as line ending, which new classes should adhere to

**Comment by WolverinDEV** [comment]
Okey well most of your points having a good sense and i will change/edit it later.
But i have some questions:

- "TS3Api l 3715, l 3756" If the port (witch has the server send) -1 then the server denied the File transfer. All other ports mean accept.
- "FileList l 70" this clause isnt  redundant. If the name equals then it dosnt matter if equals ignore case true too. But if isnt it equal then the machine checks if the boolean ignoreCase true and if its true than its check if its ignorcase equals.
- TestFileDownload open a JFrame to display the downloaded data (In my test was the data an image)
- The class `FileList` will contains some more helpfull methods. Currently its contains the Methode `getFile(String name)`

Okey now to your questions:

- multiple connections for file up/download are supported by the server.
- I dont know about any flood protection. The only protection will be the firewall of the server.
- If the size dosnt match then trow an exception.

I will submit a secon version of the merge request tomorrow with the changes.

**Comment by Firedroide** [comment]
FileList l 70:

I never said the clause was redundant, but I said that if `ignoreCase` is true, you're doing unnecessary work because `.equals` implies `.equalsIgnoreCase`. If the only difference is in capitalization, you're testing both `.equals` and `.equalsIgnoreCase`.

**Comment by WolverinDEV** [comment]
The equals methode dont imlement equalsIgnorCase()
The methode:
```java
    public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        if (anObject instanceof String) {
            String anotherString = (String)anObject;
            int n = value.length;
            if (n == anotherString.value.length) {
                char v1[] = value;
                char v2[] = anotherString.value;
                int i = 0;
                while (n-- != 0) {
                    if (v1[i] != v2[i])
                        return false;
                    i++;
                }
                return true;
            }
        }
        return false;
    }
```

**Comment by Firedroide** [comment]
Nobody has said anything about it implementing `equalsIgnoreCase`.

I was saying that `equals` **implies** `equalsIgnoreCase`.
In other words: If `equals` returns true, then `equalsIgnoreCase` must also return true.

In case `ignoreCase` is true, in most cases you're checking both `equals` **and** `equalsIgnoreCase`, which is **redundant**. You can get away with just checking **equalsIgnoreCase**.

Code says more than words in this case, I think.

    if (ignoreCase ? file.getName().equalsIgnoreCase(name) : file.getName().equals(name)) {

**Comment by WolverinDEV** [comment]
yes but java breaks if one condition true ;)

**Comment by Firedroide** [comment]
And about FileList:

We might as well also have `ClientList`, `ChannelList` and `ServerList`, which might all contain methods like getById or getByChannel or whatever.

And I'm not saying that all of these classes are necessarily a bad idea, but generally they just add tons of boilerplate code to the API that might as well not exist.

Example: Print all server query clients to standard-out

```java
ClientList clients = api.getClients();
List<Client> queryClients = clients.getServerQueryClients();
for (Client queryClient : queryClients) {
    System.out.println(queryClient);
}
```

vs

```java
List<Client> clients = api.getClients();
for (Client client : clients) {
    if (client.isServerQueryClient()) System.out.println(client);
}
```

or with Java 8

```java
List<Client> clients = api.getClients();
clients.stream()
       .filter(Client::isServerQueryClient)
       .forEach(System.out::println);
```

In my eyes, these 3 methods are all just as "good" as the other - but only method 1 requires us to have classes with hundreds of lines of code that do nothing but filter out single objects from a list.

**Comment by WolverinDEV** [comment]
Well its your API so i can remove it ;)
I think for more discussion its better to go on gitter.

**Comment by Firedroide** [comment]
> yes but java breaks if one condition true ;)

*Sound of head banging against a wall*

```java
String a = "asdf";
String b = "qwer";
boolean c = true;
return (a.equals(b) || (c && a.equalsIgnoreCase(b)));
```

Will first check if `a.equals(b)`, which returns `false`.
Then it will see that `c == true` and check `a.equalsIgnoreCase(b)`, which will also return `false`.

Do you not see how you're doing the same work twice in this case??  
If c is true, you can just call `a.equalsIgnoreCase(b)` and be done with it.
Everything else is a waste of time.

**Comment by WolverinDEV** [comment]
Okey well thats right but we live (correct me if im wrong) in the 21th century and we all have a cpu oder 1GHz (i hope :D) i think this 0.002ms dosnt mater.

**Comment by Firedroide** [comment]
No, but why should we be wasteful with our resources if we can just write our code **correctly**?
See the code I posted above on this topic. It's not more complicated than the one you're proposing and it doesn't do any redundant work.

**Comment by WolverinDEV** [comment]
Okey well i tested it on performance (i was interested).
And i think in 100000000 loops is yours better but if you take the 262ms difference and divide it with 
100000000 than you have 0.00000262 -> 2.62ns more process time with my methode. I think this isnt important but your right ;)
```java
long start;
		int loops = 100000000;
		start = System.currentTimeMillis();
		for(int i = 0;i<loops;i++)
			if("abcdefghijklmnop".equals("ponmlkjihgfedcda") || (true && "abcdefghijklmnop".equalsIgnoreCase("ponmlkjihgfedcda")));
		System.out.println("Needed time for methode 1: "+(System.currentTimeMillis()-start)+"ms");
		
		start = System.currentTimeMillis();
		for(int i = 0;i<loops;i++)
			if(true ? "abcdefghijklmnop".equalsIgnoreCase("ponmlkjihgfedcda") : "abcdefghijklmnop".equals("ponmlkjihgfedcda"));
		System.out.println("Needed time for methode 2: "+(System.currentTimeMillis()-start)+"ms");
```

Output:
`
Needed time for methode 1: 1408ms
Needed time for methode 2: 1146ms
`

**Comment by WolverinDEV** [comment]
I dont longer support my file transfare API. I dont have time to update it.
